### ISSUE TITLE
Function value for `keys` not working as expected

### ISSUE THREAD

**Issue by lucasvianav** [max397574/better-escape.nvim#6]
(also described [here](https://www.reddit.com/r/neovim/comments/ptrio7/comment/he3balm/?utm_source=reddit&utm_medium=web2x&context=3))

I tried setting a function to be ran when pressing `jk`, but it's not working for some reason. What I'd like it to do is basically send `<C-e><Esc>` when a pum is visible (to close it before escaping, as instructed by [coq](https://github.com/ms-jpq/coq_nvim/blob/coq/docs/KEYBIND.md#custom-keybindings)), and only `<Esc>` when it's not.

My config is as follows:

```lua
require('better_escape').setup({
    mapping = { 'jk' },
    timeout = 200,
    keys = function()
        print(vim.inspect(vim.fn.pumvisible()))
        return vim.fn.pumvisible() and '<C-e><Esc>' or '<Esc>'
    end,
})
```

Whenever I use `jk` to leave insert mode, `pumvisible()`'s output is `0`, even when the pum is actually visible (also doesn't matter if I have a completion option selected). 

Besides that, in this setup, it's copying the char below the cursor to the cursor position on `jk`. For example, with `|` being the cursor in insert mode and `[]` in normal mode:
```lua
print|(vim.inspect(vim.fn.pumvisible()))
return vim.fn.pumvisible() and '<C-e><Esc>' or '<Esc>'
```
*pressing `jk` copies the **n** from `return`*
```lua
print[n](vim.inspect(vim.fn.pumvisible()))
return vim.fn.pumvisible() and '<C-e><Esc>' or '<Esc>'
```

Using `keys = '<Esc>'` works fine tho. If you'd like I can make this a GH issue.

**Comment by max397574** [comment]
I think ik what happens but not why

```
						*i_CTRL-E*
CTRL-E		Insert the character which is below the cursor.
```
if you type jk you no longer get completion propositions so there is no completition menu so c-e does it's default thing (see above)
that's where the [n] comes from

that `pumvisible()`'s output is 0 doesn't make sense to me. (what completion plugin do you use)
I could imagine that it has sth todo with that you don't get completions after deleting chars but only after entering
because you enter `jk`and then they get deleted
I just don't know why `<c-e><ESC>`gets executed and not `<ESC>`only
I'll look at it next week

**Comment by max397574** [comment]
okay it's actually simple

`pumvisible()`returns 0 if there is no popupmenu

0 is evaluated as true in lua

**Comment by lucasvianav** [comment]
> pumvisible()returns 0 if there is no popupmenu

wow, I totally forgot to `== 0` it. thanks :)

> if you type jk you no longer get completion propositions so there is no completition menu so c-e does it's default thing (see above)

what do you mean? what I'm trying to solve is that when I have a completion option selected, if I just pres `jk` it'll confirm the completion before escaping insert mode. If i do `ino <silent><expr> jk   pumvisible() ? "\<C-e><Esc>" : "\<Esc>"` it works --- just escapes insert mode without confirming the completion.

I'm using [coq_nvim](https://github.com/ms-jpq/coq_nvim) btw

**Comment by lucasvianav** [comment]
> pumvisible()returns 0 if there is no popupmenu

wow, I totally forgot to `== 0` it. thanks :)

> if you type jk you no longer get completion propositions so there is no completition menu so c-e does it's default thing (see above)

what do you mean? what I'm trying to solve is that when I have a completion option selected, if I just pres `jk` it'll confirm the completion before escaping insert mode.

![see this gif](https://user-images.githubusercontent.com/23108450/134824182-7e21c1e5-ef71-4055-b90e-8704a88fe5d8.gif)

If i do `ino <silent><expr> jk   pumvisible() ? "\<C-e><Esc>" : "\<Esc>"` it works --- just escapes insert mode without confirming the completion.

I'm using [coq_nvim](https://github.com/ms-jpq/coq_nvim) btw

**Comment by lucasvianav** [comment]
I've added the `== 0` now. Sometimes `pumvisible()` will return 1, other times 0, regardless if the pum is actually visible or not. I tried changing my config to:
```lua
    keys = function()
        local selected = vim.fn.complete_info({ 'selected' }).selected
        return selected ~= -1 and '<C-e><Esc>' or '<Esc>'
    end,
```

But `selected` is always -1, even when I do have a completion option enabled

**Comment by max397574** [comment]
Actually the plugin (with the first config you sent) checks whether there is a pum after deleting the j and k

so it's the same like when you slowly type jk<BS><BS> 
can you check if there is a pum then?

**Comment by lucasvianav** [comment]
Yeah, when mapping `jk` directly the pum is still visible when the expression is evaluated. Look at the gifs below.

![Peek 2021-09-27 10-51](https://user-images.githubusercontent.com/23108450/134923765-4c4d44fd-ed53-494d-8ca6-2dc59aee6203.gif)

![Peek 2021-09-27 10-59](https://user-images.githubusercontent.com/23108450/134923770-ef9c32e5-1a69-47db-bb38-8cd8c6e9b4a7.gif)

Here's the function I've used:
```
function! Test() abort
    echom complete_info()
    return pumvisible() ? "\<C-e>\<Esc>" : "\<Esc>"
endfunction

" ino <silent><expr> jk pumvisible() ? "\<C-e><Esc>" : "\<Esc>"
ino <expr> jk Test()
```

Would it be bad to run the `keys` function **before** deleting `jk`? Or maybe expose an option for the user to choose.

**Comment by max397574** [comment]
This would be possible
But I have to try if it doesn't break anything first
I'll ping you as soon as it's on the dev branch so you can test

**Comment by max397574** [comment]
It's not possible to run the function before. Because the function could be something that should be done after jk are deleted. I guess you could just use `<ESC>`and not a function and it should work for you.

**Comment by lucasvianav** [comment]
Wouldn't it be possible to expose the option of running the function either before or after?

**Comment by lucasvianav** [comment]
> I think it would be. I'll look at it next week.

Alright, thanks :)

**Comment by max397574** [comment]
@lucasvianav 
added it on the dev branch

you can try
`keys_before_delete = true,`
in the setttings

**Comment by lucasvianav** [comment]
thanks @max397574! I was about to test it, but I think there's a typo
```
Error detected while processing InsertCharPre Autocommands for "*":
E5108: Error executing lua ...pack/packer/opt/better-escape.nvim/lua/better_escape.lua:53: attempt to index global 'settigns' (a nil value)
Press ENTER or type command to continue
```

**Comment by lucasvianav** [comment]
I fixed the typo and tried it, but feeding `<BS><BS>` **after** exiting to normal mode goes back 2 chars (even sending me some lines up if I'm at the start of line). I'm not able to record a GIF right now tho

I also tried running a function both before and after `jk` is deleted (see below) -- the idea was to feed `<C-e>` before if needed and only leave insert mode after deleting `jk`. But it also doesn't work, for some reason the function isn't always able to detect that the pum is visible nor that there's an option selected, I couldn't figure out why.

![image](https://user-images.githubusercontent.com/23108450/136669455-5d547435-c769-433f-b897-1fd1a21fd48b.png)
![image](https://user-images.githubusercontent.com/23108450/136669340-4511929d-62f1-4317-baa9-7632de13d424.png)

**Comment by lucasvianav** [comment]
For me it doesn't, unfortunately. Are you using coq?

**Comment by max397574** [comment]
In the gif you sent

> ![see this gif](https://user-images.githubusercontent.com/23108450/134824182-7e21c1e5-ef71-4055-b90e-8704a88fe5d8.gif)

do you use a function there or just the default settings?
@lucasvianav

**Comment by max397574** [comment]
so it basically expands some time, some times it deletes a character

imo this can't be an error from better-escape.nvim because it does only feed <BS><BS><ESC>
and this shouldn't expand or anything

**Comment by lucasvianav** [comment]
yeah, I'm with you on this, I don't think it's a better-escape.nvim problem. 

it's just a particularity about coq_nvim that when escaping with a completion option selected, you have to feed '<C-e>' before '<Esc>', and it'd be nice if better-escape could support it. unfortunately I don't think it's possible, although I don't know exactly what the incompatibility is

regardless, thanks for all the help and the nice plugin :)
